server/src/utils.ts

```diff
@@ -95,3 +96,42 @@ export function urlDecodeSpecialPathChars(
   req.url = urlFormat(url);
   next();
 }
+
+/**
+ * Get the complete host for a request. This differs from `request.hostname`
+ * by including the port.
+ *
+ * The code here is mostly stolen from Express 5.x's `request.host`; in
+ * Express 4.x, this property is an alias for `hostname` and has no port.
+ */
+export function getRequestHost(request: Request): string {
+  const trust = request.app.get("trust proxy fn");
+  let val = request.get("X-Forwarded-Host");
+
+  if (!val || !trust(request.socket.remoteAddress, 0)) {
+    val = request.get("Host");
+  } else if (val.indexOf(",") !== -1) {
+    // Note: X-Forwarded-Host is normally only ever a
+    //       single value, but this is to be safe.
+    val = val.substring(0, val.indexOf(",")).trimEnd();
+  }
+
+  return val || undefined;
+}
+
+/**
+ * Get the configured host URL (a.k.a. origin) for the server, or if none is
+ * set, the origin of a given request.
+ */
+export function getHostUrl(request?: Request): string {
+  let hostUrl = getPrimaryHostUrl();
```

Comment:
Yeah, I did not feel super happy with this either. I could probably do better, but also felt like I got myself in a sticky spot from my earlier decision to name the environment/config variable `HOST_URL`, since it’s really an origin and not a host. (Maybe it should have been `ORIGIN` or some variation thereof? `PRIMARY_ORIGIN`/`CANONICAL_ORIGIN`/`DEFAULT_ORIGIN`? Maybe I should have made it just a host and not an origin? Ugh.)

> Did you consider splitting the responsibility here the other way…

No, but mainly because I was stuck on an old idea of how I thought this was going to be used, which never panned out. I think you’re on the right track here; will try re-doing it that way.

Reply:
If you're down to go all the way to changing the name of the environment var, I would for sure vote for it just being a hostname rather than a URL. In nginx the concept is called a ["primary server name"](http://nginx.org/en/docs/http/ngx_http_core_module.html#server_name) and Caddy calls it a ["site address"](https://caddyserver.com/docs/caddyfile/concepts), if those serve as inspiration for how to name it.